File: challenges/2023/02.js
// Part 1
const solvePuzzle1 = (input) => {
  const limits = { red: 12, green: 13, blue: 14 };

  const games = input.split('\n').map((line, idx) => {
    const rounds = line.split(': ')[1];

    return rounds
      .replaceAll(' ', '')
      .split(';')
      .map((round) => {
        const colorsAndNumbers = {
          red: 0,
          green: 0,
          blue: 0,
          gameNo: idx + 1,
        };

        round.split(',').forEach((numAndColor) => {
          const number = numAndColor.match(/^\d+/)[0];
          const color = numAndColor.slice(number.length);
          colorsAndNumbers[color] += Number(number);
        });

        return colorsAndNumbers;
      });
  });

  const legalGames = games.filter((rounds) => {
    return rounds.every((game) => game.red <= limits.red && game.green <= limits.green && game.blue <= limits.blue);
  });

  return legalGames.reduce((acc, rounds) => rounds[0].gameNo + acc, 0);
};

// Part 2
const solvePuzzle2 = (input) => {
  const games = input.split('\n').map((line, idx) => {
    const rounds = line.split(': ')[1];

    return rounds
      .replaceAll(' ', '')
      .split(';')
      .map((round) => {
        const colorsAndNumbers = {
          red: 0,
          green: 0,
          blue: 0,
          gameNo: idx + 1,
        };

        round.split(',').forEach((numAndColor) => {
          const number = numAndColor.match(/^\d+/)[0];
          const color = numAndColor.slice(number.length);
          colorsAndNumbers[color] += Number(number);
        });

        return colorsAndNumbers;
      });
  });

  const minimumRequiredBallsForGames = games.map((rounds) => {
    const rgbMinimums = [0, 0, 0];

    rounds.forEach((round) => {
      rgbMinimums[0] = Math.max(rgbMinimums[0], round.red);
      rgbMinimums[1] = Math.max(rgbMinimums[1], round.green);
      rgbMinimums[2] = Math.max(rgbMinimums[2], round.blue);
    });

    return rgbMinimums;
  });

  return minimumRequiredBallsForGames.reduce((acc, rounds) => rounds[0] * rounds[1] * rounds[2] + acc, 0);
};

require(__dirname + '/../../utils/test.js').test(__filename, __dirname, solvePuzzle1, '1');
require(__dirname + '/../../utils/test.js').test(__filename, __dirname, solvePuzzle2, '2');

File: challenges/2023/02.exs
Code.compile_file(Path.join([__ENV__.file], ["../../../utils/utils.exs"]))

defmodule Challenge do
  @moduledoc false

  def first_result(input) do
    input
    |> parse_games()
    |> Enum.filter(&within_limits?/1)
    |> Enum.reduce(0, &(Enum.at(&1, 0)["game_no"] + &2))
  end

  def second_result(input) do
    input
    |> parse_games()
    |> Enum.reduce(0, fn rounds, acc ->
      [r, g, b] =
        Enum.reduce(rounds, [0, 0, 0], fn round, [r, g, b] ->
          [max(round["red"], r), max(round["green"], g), max(round["blue"], b)]
        end)

      r * g * b + acc
    end)
  end

  defp within_limits?(rounds) when is_list(rounds), do: Enum.all?(rounds, &within_limits?/1)
  defp within_limits?(r) when is_map(r), do: r["red"] <= 12 && r["green"] <= 13 && r["blue"] <= 14

  defp parse_games(input) do
    input
    |> Enum.with_index(1)
    |> Enum.map(fn {game, game_no} ->
      game
      |> String.split(": ")
      |> List.last()
      |> String.replace(" ", "")
      |> String.split(";")
      |> Enum.map(fn round ->
        colors_and_numbers = %{"red" => 0, "green" => 0, "blue" => 0, "game_no" => game_no}

        round
        |> String.split(",")
        |> Enum.reduce(colors_and_numbers, fn round, acc ->
          ["", number_string, color] = Regex.split(~r/\d+/, round, include_captures: true)
          Map.put(acc, color, String.to_integer(number_string))
        end)
      end)
    end)
  end
end

Utils.run(&Challenge.first_result/1, __ENV__.file, "Q1")
Utils.run(&Challenge.second_result/1, __ENV__.file, "Q2")
